File: Python/0527/test2.py
# class Employee5:
#     empCount = 0
#     def __init__(self,name, salary):
#         self.name = name
#         self.salary = salary
#         Employee5.empCount += 1
#     def displayCount(self):
#         print("Employee %d" % Employee5.empCount)

#     def displayEmployee(self):
#         print("name : ",self.name, ", salary : ", self.salary)

# emp1 = Employee5("이순신", 2000)
# emp1.displayCount()
# emp1.displayEmployee()
# emp2 = Employee5("강감찬", 3000)
# emp2.displayCount()
# emp2.displayEmployee()
# print(Employee5.__dict__)

class Animal:
    def berath(self):
        print("Animal breath")
    def eat(self, name ="강아지", age = 20):
        self.name = name
        self.age = age
class Dog(Animal):
    def bark(self):
        print("Dog barking")
    def breath(self): ##overriding
        print("Dog breathing")

d = Dog()
d.bark()
d.berath()
a = Animal()
a.eat(name = "강아지2")

# class MyData:
#     __c = 10
#     def __init__(self,a,b,c):
#         self.a = a
#         self.b = b
#         self.__c = c
#     def displayData(self):
#         print(self.a, self.b)
#     def  __displayData2(self):
#         print("Nothing")
#     def __del__(self):
#         print("Delete")

# MyData = MyData(1,2,3)
# print(MyData.a)

# class Calc1:
#     def Add(self,a,b):
#         return a + b
# class Calc2:
#     def multiply(self, a,b):
#         return a*b
# class Calc3(Calc1, Calc2):
#     def divide(self,a,b):
#         return a/b

# c = Calc3()
# print(c.Add(1,2))
# print(c.multiply(3,2))
# print(c.divide(4,2))

File: JavaTest_0416/test0416/Test2.java
package com.test0416;
//class가 interface를 상속하는 행위는 '구현'
interface Animal {
	void breath();//추상메소드(몸체가 비어있음)
	
}

class Dog implements Animal {
	public void breath() {//구현메소드
		System.out.println("Dog breath");
	}
	public void bark() {
		System.out.println("Dog Bark");
	}
}

class Cat implements Animal {
	public void breath() {
		System.out.println("Cat breath");
	}
	public void meow() { 
		System.out.println("Cat meow");
	}
}


public class Test2 {
	static void test(Animal a) {
		a.breath();
	}

	public static void main(String[] args) {
//		Animal a = new Animal(); class에 interface와 똑같은 메소드가 구현되지 않으면 객체화 되지 않음.
		Dog d = new Dog(); 
		Animal a = d; //이런식으로 class먼저 호출한 다음에 해야함.
		a.breath();
		
		Animal c = new Cat(); //위의 두줄의 식을 한번에 적을 때
		c.breath();
		
		test(new Dog());
		test(new Cat());
	}

}
